### ISSUE TITLE
`tapResponse` swallowing errors happening inside the `tapResponse`

### ISSUE THREAD

**Issue by maxime1992** [ngrx/platform#3431]
### Minimal reproduction of the bug/regression with instructions

```
    // not working as expected
    of(null)
      .pipe(
        tapResponse(
          () => {
            throw new Error(
              `This error will be swallowed and you'll have a hard time debugging`
            );
          },
          (error) => console.log('Some error happened', error)
        )
      )
      .subscribe();
```

That error :point_up: is silently caught which was painful to debug in my case. Here's a live repro on stackblitz

https://stackblitz.com/edit/angular-ivy-7jhxew?file=src%2Fapp%2Fapp.component.ts

### Minimal reproduction of the bug/regression with instructions

To have an error thrown correctly.

### Versions of NgRx, Angular, Node, affected browser(s) and operating system(s)

![image](https://user-images.githubusercontent.com/4950209/170252955-feb4cc45-ecbe-4e56-aee9-989a0eb7e0f3.png)

### Other information

This very much looks like an issue I had with `ngrxPush` pipe here: https://github.com/ngrx/platform/issues/3100

### I would be willing to submit a PR to fix this issue

- [X] Yes
- [ ] No

**Comment by maxime1992** [comment]
Not sure what's best here between
- not handling an error that'd happen in the `tapResponse` `next` callback so that it's broadcasted all the way up
- handle the error happening within the `tapResponse` `next` callback within the catch error of that custom operator

**Comment by markostanimirovic** [comment]
@maxime1992

The `tapResponse` is used to keep the effect active if an error is thrown. Because of that, `tapResponse` swallows the thrown error. The `error` callback is not executed, because the error that is thrown within the `next` callback of `tap` operator will not be caught by its `error` callback:

```ts
**tap**({
  next() {
    throw new Error('ERROR!');
  },
  error(err) {
    // ERROR! won't be caught here
  }
})
```

The solution could be to move the `error` callback execution from `tap` to `catchError`:

```ts
function tapResponse(next, error, complete?) {
  return pipe(
    tap({ next, complete }),
    catchError((err) => {
      error(err);
      return EMPTY;
    });
  );
}
```

cc @alex-okrushko 

---

For now, you can use `tap` + `catchError`:

```ts
.pipe(
  tap(() => {
    throw new Error('ERROR!');
  }),
  catchError((err) => {
    // handle error here

    return EMPTY;
   })
)
```

**Comment by timdeschryver** [comment]
> The solution could be to move the error callback execution from tap to catchError

I would find that weird. For me, or how I think about it, the `error` callback is to handle errors that are thrown while doing the request, not for errors in `next`.

What about adding a config, similar to the [EffectConfig](https://github.com/ngrx/platform/blob/master/modules/effects/src/models.ts#L13) to bypass the default behavior?

**Comment by maxime1992** [comment]
> What about adding a config, similar to the [EffectConfig](https://github.com/ngrx/platform/blob/master/modules/effects/src/models.ts#L13) to bypass the default behavior?

I don't think this would be good enough. Unless this is pretty much the first thing first first thing in the documentation, someone could miss it and the frustration to understand what's going on with a silently swallowed error can be really high. I do believe there's a need for a default behavior where you can handle this. I can't think of a case where someone would let its stream error without being warned at least

**Comment by spierala** [comment]
Probably a super naive solution, still I want to share it:

Try catch around the nextFn...

```
export function tapResponse<T, E = unknown>(
    nextFn: (next: T) => void,
    errorFn: (error: E) => void,
    completeFn?: () => void
): (source: Observable<T>) => Observable<T> {
    return (source) => {
        return source.pipe(
            tap({
                next: (v) => {
                    try {
                        nextFn(v);
                    } catch (error) {
                        console.error('An error occurred in the next callback', error);
                    }
                },
                error: errorFn,
                complete: completeFn,
            }),
            catchError(() => EMPTY)
        );
    };
}
```

**Comment by spierala** [comment]
Fixed it with try catch in MiniRx :) https://github.com/spierala/mini-rx-store/pull/117/files

**Comment by alvipeo** [comment]
Just spent hours finding this problem. Here's the working example - https://stackblitz.com/edit/angular-ivy-nxf2la?file=src/app/app.component.html.

**Comment by balajibadly** [comment]
> Just spent hours finding this problem. Here's the example showing the problem - https://stackblitz.com/edit/angular-ivy-nxf2la.

what is the solution for this? any work around @markostanimirovic?

**Comment by alex-okrushko** [comment]
Even through there is an expectation that `next` callback doesn't throw an error, we can adjust to catch errors in the `next` as well, I don't think there's a problem.

The question that I have is would you expect an error in the `error` callback to be handled somehow as well?

**Comment by alvipeo** [comment]
In my opnion, whatever happens in both - no exception must be swallowed. If it throws in `next` => `error` should be fired. If it's thrown in `error` => console.error by default. But add the option to override this.

**Comment by markostanimirovic** [comment]
Closed by #3533

**Comment by joelcoxokc** [comment]
Hello, I am still experiencing this issue in ngrx/operators@19.0.1
